src/types.rs

```diff
@@ -57,108 +57,517 @@ use crate::{
+/// current node, and let the traversal take care of the plumbing of recursion and reconstruction.
+///
+/// ## Type parameters
+///
+/// Here, `TypeF` also takes two additional type parameters for the recursive unfolding of record
+/// rows and enum rows. We have other, distinct recursive subcomponents (or subtrees) in our
+/// complete definition, but the approach is unchanged.
 #[derive(Clone, PartialEq, Debug)]
-pub enum AbsType<Ty> {
-    /// The dynamic unitype, affected to values which type is not statically known or enforced.
-    Dyn(),
+pub enum TypeF<Ty, RRows, ERows> {
+    /// The dynamic type, or unitype. Affected to values which actual type is not statically known
+    /// or checked.
+    Dyn,
     /// A floating point number.
-    Num(),
+    Num,
     /// A boolean.
-    Bool(),
+    Bool,
     /// A string literal.
-    Str(),
+    Str,
     /// A symbol.
     ///
     /// See [`crate::term::Term::Sealed`].
-    Sym(),
+    Sym,
     /// A type created from a user-defined contract.
     Flat(RichTerm),
     /// A function.
     Arrow(Ty, Ty),
     /// A type variable.
     Var(Ident),
     /// A forall binder.
-    Forall(Ident, Ty),
-
-    /// An empty row, terminating a row type.
-    RowEmpty(),
-    /// A row type.
-    RowExtend(
-        Ident,
-        Option<Ty>, /* Type of the field, or None for enums */
-        Ty,         /* Tail (another row) */
-    ),
-    /// An enum type, wrapping a row type for enums.
-    Enum(Ty /* Row */),
-    /// A record type, wrapping a row type for records.
-    Record(Ty /* Row */),
+    Forall {
+        var: Ident,
+        var_kind: VarKind,
+        body: Ty,
+    },
+
+    /// An enum type, composed of a sequence of enum rows.
+    Enum(ERows),
+    /// A record type, composed of a sequence of record rows.
+    Record(RRows),
     /// A dictionary type.
-    // Dict will only have a default type, this is simpler for now, I don't think we lose much
     Dict(Ty),
     /// A parametrized array.
     Array(Ty),
     /// A type wildcard, wrapping an ID unique within a given file.
     Wildcard(usize),
 }
 
-impl<Ty> AbsType<Ty> {
-    pub fn try_map<To, F, E>(self, mut f: F) -> Result<AbsType<To>, E>
+// Concrete, recursive definition of Nickel types from the generic `XxxF` definitions. This is
+// "tying" the note. We have to put `Box` in the appropriate positions (otherwise, Rust will
+// complain that the type has an infinite size), but also avoid putting more than necessary.
+//
+// For example, `RecordRows` contains a `RecordRow`. The latter doesn't need to be boxed, because a
+// `RecordRow` itself potentially contains occurrences of `Types` and `RecordRows`, which need to
+// be boxed. Hence, we don't need to additionally box `RecordRow`.
+
+/// Concrete, recursive definition for enum rows.
+#[derive(Clone, PartialEq, Debug)]
+pub struct EnumRows(pub EnumRowsF<Box<EnumRows>>);
+/// Concrete, recursive definition for a record row.
+pub type RecordRow = RecordRowF<Box<Types>>;
```

Comment:
Because the definition is not recursive (at least, not directly). If I could, I would have defined everything as type aliases, but you can't define recursive type aliases (for obvious reasons, the first being that the size of such a type would be infinite).

That being said, it does bring non-uniformity: for some reason, one type doesn't need a constructor. Do you think it would be less confusing to define it as a newtype, even if it doesn't strictly need to?

Reply:
Reading the code that actually uses `RecordRow` didn't feel confusing to me, at least not for that reason. I think the type system makes and LSP completion makes this mostly a discussion about aesthetics, so either way should be fine.

I'd leave it up to you whether you want some extra uniformity or whether we should just get on with it and leave it as a type alias.